pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,293 @@
+	if err != nil {
+		klog.ErrorS(err, "Invalid PodLifeTime parameters")
+		return
+	}
+
+	nodeMap := make(map[string]*v1.Node, len(nodes))
+	for _, node := range nodes {
+		nodeMap[node.Name] = node
+	}
+	evictable := podEvictor.Evictable(evictions.WithPriorityThreshold(thresholdPriority))
+
+	// 1. for each namespace for which there is Topology Constraint
+	// 2. for each TopologySpreadyConstraint in that namespace
+	//  { find all evictable pods in that namespace
+	//  { 3. for each evictable pod in that namespace
+	// 4. If the pod matches this TopologySpreadConstraint LabelSelector
+	// 5. If the pod nodeName is present in the nodeMap
+	// 6. create a topoPair with key as this TopologySpreadConstraint.TopologyKey and value as this pod's Node Label Value for this TopologyKey
+	// 7. add the pod with key as this topoPair
+	// 8. find the min number of pods in any topoPair for this topologyKey
+	// iterate through all topoPairs for this topologyKey and diff currentPods -minPods <=maxSkew
+	// if diff > maxSkew, add this pod in the current bucket for eviction
+
+	// First record all of the constraints by namespace
+	namespaces, err := client.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Couldn't list namespaces")
+		return
+	}
+	podsForEviction := make(map[*v1.Pod]struct{})
+	// 1. for each namespace...
+	for _, namespace := range namespaces.Items {
+		if (!includedNamespaces.Has(namespace.Name) || excludedNamespaces.Has(namespace.Name)) && (includedNamespaces.Len()+excludedNamespaces.Len() > 0) {
+			continue
+		}
+		namespacePods, err := client.CoreV1().Pods(namespace.Name).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			klog.ErrorS(err, "Couldn't list pods in namespace", "namespace", namespace)
+			return
+		}
+
+		// ...where there is a topology constraint
+		namespaceTopologySpreadConstrainPods := make([]v1.Pod, 0, len(namespacePods.Items))
+		for _, pod := range namespacePods.Items {
+			if len(pod.Spec.TopologySpreadConstraints) > 0 {
+				namespaceTopologySpreadConstrainPods = append(namespaceTopologySpreadConstrainPods, pod)
+			}
+		}
+		if len(namespaceTopologySpreadConstrainPods) == 0 {
+			return
+		}
+
+		// 2. for each topologySpreadConstraint in that namespace
+		for _, constrainedPod := range namespaceTopologySpreadConstrainPods {
+			// find pods to evict one constraint at a time
+			for _, constraint := range constrainedPod.Spec.TopologySpreadConstraints {
+				// Only deal with hard topology constraints
+				// TODO(@damemi): add support for soft constraints
+				if constraint.WhenUnsatisfiable != v1.DoNotSchedule {
+					continue
+				}
+				constraintTopologies := make(map[topologyPair][]*v1.Pod)
+				// pre-populate the topologyPair map with all the topologies available from the nodeMap
+				// (we can't just build it from existing pods' nodes because a topology may have 0 pods)
+				for _, node := range nodeMap {
```

Comment:
> This `for` loop is expensive. I'd suggest putting it in an outer loop right after we have calculated all constraints within the current namespace.

The problem is this loop is meant to identify any topologyPairs/domains that have 0 pods in them. The only way for us to know that a label on a pair is a topologyKey (and not just a label) is to check against each constraint's key. So even if it's moved outside the loop, it will still end up being O(M*N) where we have to check each node for each constraint.

> And, IMO even we should move it to the most outer loop after we have all constraints over all eligible namespaces - i.e., constraintTopologies is a map keyed with namespace, and values with `map[topologyPair][]*v1.Pod` - which would be the most efficient way.

I'm not sure the efficiency that's gained here, as we will still be looping over each namespace->constraint

> BTW: you can move the logic of building nodeMap below calculating constraints, as if len(constraints) == 0, we can return earlier.

This loop is ranging over `constraints`, so if `len(constraints) == 0` we don't have to worry about that